source/server/admin/ungrouped_stats_request.cc

```diff
@@ -0,0 +1,108 @@
+#include "source/server/admin/ungrouped_stats_request.h"
+
+namespace Envoy {
+namespace Server {
+
+UngroupedStatsRequest::UngroupedStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                             UrlHandlerFn url_handler_fn)
+    : StatsRequest(stats, params, url_handler_fn) {
+
+  phases_ = {Phase{PhaseName::TextReadouts, "Text Readouts"},
+             Phase{PhaseName::CountersAndGauges, "Counters and Gauges"},
+             Phase{PhaseName::Histograms, "Histograms"}};
+
+  switch (params_.type_) {
+  case StatsType::TextReadouts:
+  case StatsType::All:
+    phase_index_ = 0;
+    break;
+  case StatsType::Counters:
+  case StatsType::Gauges:
+    phase_index_ = 1;
+    break;
+  case StatsType::Histograms:
+    phase_index_ = 2;
+    break;
+  }
+}
+
+template <class StatType> Stats::IterateFn<StatType> UngroupedStatsRequest::saveMatchingStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if used
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // Capture the name if we did not early-exit due to used_only -- we'll use
+    // the name for both filtering and for capturing the stat in the map.
+    // stat->name() takes a symbol table lock and builds a string, so we only
+    // want to call it once.
+    //
+    std::string name = stat->name();
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(name, *params_.filter_)) {
+        return true;
+      }
+    } else if (params_.re2_filter_ != nullptr &&
+               !re2::RE2::PartialMatch(name, *params_.re2_filter_)) {
+      return true;
+    }
```

Comment:
This is the same as the prometheus filtering. Can it be pulled out to a utility/helper?

Reply:
Very similar but not exactly the same: the prometheus logic avoids a `stat->name()` unless there is a filter. The original code had a comment explaining this, I will add it back.